phper/src/builder/data_type.rs

```diff
@@ -0,0 +1,168 @@
+use crate::builder::describe::Describe;
+
+pub enum ArrayKeyType {
+    String,
+    Int,
+    Mixed,
+}
+
+pub enum AtomicType {
+    Bool,
+    Int,
+    Float,
+    String,
+    NonEmptyString,
+    IntRange(i64, i64),
+    ClassString,
+    ClassStringOf(String),
+    TraitString,
+    EnumString,
+    CallableString,
+    NumericString,
+    LiteralString,
+    LiteralInt,
+    Numeric,
+    Scalar,
+    Object,
+    ObjectWithProperties(Vec<(String, AtomicType)>),
+    ObjectOf(String),
+    List(Box<AtomicType>),
+    Array(Box<(ArrayKeyType, AtomicType)>),
+    NonEmptyList(Box<AtomicType>),
+    NonEmptyArray(Box<(ArrayKeyType, AtomicType)>),
+    Shape(Vec<(ArrayKeyType, AtomicType)>),
+    Iterable(Box<(ArrayKeyType, AtomicType)>),
+    CallableArray,
+    Callable,
```

Comment:
I'm not sure if we should support callable types, i believe it would be better if had only `Closure(Vec<Argument>, AtomicReturnType)`.

wdyt @jmjoy

Reply:
also, should we have seperate `Closure` and `StaticClosure`?

Reply:
Callable is not only Function, but also String or Array.

I'm not very clear about the mechanism of PHP 8's registration function. If the parameter type declaration has a callable type, it should be required.

Reply:
`callable` is the old PHP way of saying `Closure`, currently, you can create a closure from callable, or make a closure from functions/methods, e.g:

| type | callable | Closure |
|------|----------|---------|
| function | `$a = 'name'` | `$a = name(...);` |
| method | `$a = [$obj, 'name']` | `$a = $obj->name(...);` |
| static method| `$a = ['Foo', 'name']` or `$a = 'Foo::name'` | `$a = Foo::name(...);` |

my point is we don't really need to support `callable`, people shouldn't be using it anymore, however, if you already have a reference to a `callable`, you can turn it into a `Closure` with:

`$a = Closure::fromCallable($callable)`

anonymous functions in PHP are already closures:

```php
$a = function() { };

assert($a instanceof Closure);
```

Reply:
less types to support = less work for us :) forcing users to adopt newer types/functionalities is a good thing.

Reply:
OK, callable can be not supported.